internal/beater/beater.go

```diff
@@ -827,26 +821,49 @@ func newSourcemapFetcher(
 			}
 		}
 
-		return sourcemap.NewFleetFetcher(
+		ff, err := sourcemap.NewFleetFetcher(
 			&client,
 			fleetCfg.AccessAPIKey,
 			fleetServerURLs,
 			artifactRefs,
 		)
+		if err != nil {
+			return nil, err
+		}
+
+		cachingFetcher, err := sourcemap.NewCachingFetcher(ff, cfg.Cache.Expiration)
+		if err != nil {
+			return nil, err
+		}
+
+		return cachingFetcher, nil
 	}
 
 	// For standalone, we query both Kibana and Elasticsearch for backwards compatibility.
 	var chained sourcemap.ChainedFetcher
 	if kibanaClient != nil {
-		chained = append(chained, sourcemap.NewKibanaFetcher(kibanaClient))
+		cachingFetcher, err := sourcemap.NewCachingFetcher(sourcemap.NewKibanaFetcher(kibanaClient), cfg.Cache.Expiration)
+		if err != nil {
+			return nil, err
+		}
+
+		chained = append(chained, cachingFetcher)
 	}
 	esClient, err := newElasticsearchClient(cfg.ESConfig)
 	if err != nil {
 		return nil, err
 	}
 	index := strings.ReplaceAll(cfg.IndexPattern, "%{[observer.version]}", version.Version)
 	esFetcher := sourcemap.NewElasticsearchFetcher(esClient, index)
-	chained = append(chained, esFetcher)
+	cachingFetcher, err := sourcemap.NewCachingFetcher(esFetcher, cfg.Cache.Expiration)
+	if err != nil {
+		return nil, err
+	}
+	metadataCachingFetcher := sourcemap.NewMetadataCachingFetcher(esClient, cachingFetcher, index)
```

Comment:
**question**: 

> A configuration section (similar to the [config for agent config](https://github.com/elastic/apm-server/pull/8193/files#diff-fbadce52f216b92bda4dab62b076f334502b170af5e51bf3e442e2033b4a4e9dR201-R223)) needs to be added for fetching from Elasticsearch. Querying these ES indices will require additional privileges.

I'm not sure I get this part. The ES fetcher is using the configured es client and reusing the old indices. The metadata fetcher is sharing the same client and querying the same index.
Why does it need additional privileges ? 
If it needs additional privileges does that mean that the current ES fetcher is not working ?

Reply:
I've added a section to the agent config to inject the api key based on the open kibana PR